src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_hwctx.c

```diff
@@ -0,0 +1,366 @@
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+	struct kds_client_cu_info cu_info = {};
+	int ret = 0;
+
+        mutex_lock(&client->lock);
+
+        hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_args->hw_context);
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+                ret = -EINVAL;
+		goto out;
+        }
+
+	/* Bitstream is locked. No one could load a new one
+	 * until this HW context is closed.
+	 */
+	ret = xocl_cu_ctx_to_info(xdev, drm_cu_args, hw_ctx, &cu_info);
+	if (ret) {
+		userpf_err(xdev, "No valid CU ctx found for this HW context");
+		goto out;
+	}
+
+	/* Allocate a free CU context for the given CU index */
+	cu_ctx = kds_alloc_cu_hw_ctx(client, hw_ctx, &cu_info);
+	if (!cu_ctx) {
+		userpf_err(xdev, "Allocation of CU context failed");
+                ret = -EINVAL;
+		goto out;
+	}
+
+        ret = kds_add_context(&XDEV(xdev)->kds, client, cu_ctx);
+        if (ret) {
+                kds_free_cu_ctx(client, cu_ctx);
+		goto out;
+	}
+
+	// Return the encoded cu index along with cu domain
+	drm_cu_args->cu_index = set_domain(cu_ctx->cu_domain,
+					   cu_ctx->cu_idx);
+out:
+	mutex_unlock(&client->lock);
+        return ret;
+}
+
+int xocl_close_cu_context(struct xocl_dev *xdev, struct drm_file *filp,
+		struct drm_xocl_close_cu_ctx *drm_cu_args)
+{
+	struct kds_client *client = filp->driver_priv;
+	struct kds_client_hw_ctx *hw_ctx = NULL;
+	struct kds_client_cu_ctx *cu_ctx = NULL;
+        struct kds_client_cu_info cu_info = {}; 
+        int ret = 0;
+
+        mutex_lock(&client->lock);
+
+        hw_ctx = kds_get_hw_ctx_by_id(client, drm_cu_args->hw_context);
+        if (!hw_ctx) {
+                userpf_err(xdev, "No valid HW context is open");
+                ret = -EINVAL;
+                goto out;
+        }
+
+	xocl_close_cu_ctx_to_info(drm_cu_args, &cu_info);
+	
+	/* Get the corresponding CU Context */ 
+        cu_ctx = kds_get_cu_hw_ctx(client, hw_ctx, &cu_info);
+        if (!cu_ctx) {
+                userpf_err(xdev, "No CU context is open");
+                ret = -EINVAL;
+                goto out;
+        }
+
+        ret = kds_del_context(&XDEV(xdev)->kds, client, cu_ctx);
+        if (ret)
+                goto out;
+
+        ret = kds_free_cu_ctx(client, cu_ctx);
+
+out:
+        mutex_unlock(&client->lock);
+	return ret;
+}
+
+int xocl_hw_ctx_command(struct xocl_dev *xdev, void *data,
+			      struct drm_file *filp)
+{
+	struct drm_xocl_hw_ctx_execbuf *args = data;
+	struct drm_xocl_execbuf legacy_args = {};
+
+	/* Update the legacy structure with the appropiate value */
+	legacy_args.ctx_id = args->hw_ctx_id;
+	legacy_args.exec_bo_handle = args->exec_bo_handle;
+
+	return xocl_command_ioctl(xdev, &legacy_args, filp, false);
+}
+
+/*
+ * Open a context (only shared supported today) on a CU under the given hw_context.
+ * Return the acquired cu index for further reference.
+ */
+int xocl_open_cu_ctx_ioctl(struct drm_device *dev, void *data,
```

Comment:
All of the ioctl calls are defined in xocl_ioctl.c. Can we keep this rules until we plan to change it?

Reply:
Sure. Thanks